src/java/org/apache/cassandra/schema/MigrationCoordinator.java

```diff
@@ -369,22 +418,31 @@ private synchronized boolean shouldApplySchemaFor(VersionInfo info)
 
     synchronized Future<Void> reportEndpointVersion(InetAddressAndPort endpoint, UUID version)
     {
+        logger.debug("Reported schema {} at endpoint {}", version, endpoint);
         if (ignoredEndpoints.contains(endpoint) || IGNORED_VERSIONS.contains(version))
         {
             endpointVersions.remove(endpoint);
             removeEndpointFromVersion(endpoint, null);
+            logger.debug("Discarding endpoint {} or schema {} because either endpoint or schema version were marked as ignored", endpoint, version);
             return FINISHED_FUTURE;
         }
 
         UUID current = endpointVersions.put(endpoint, version);
         if (current != null && current.equals(version))
+        {
+            logger.trace("Skipping report of schema {} from {} because we already know that", version, endpoint);
             return FINISHED_FUTURE;
+        }
 
         VersionInfo info = versionInfo.computeIfAbsent(version, VersionInfo::new);
         if (Objects.equals(schemaVersion.get(), version))
+        {
             info.markReceived();
+            logger.trace("Schema {} from {} has been marked as recevied because it is equal the local schema", version, endpoint);
+        }
         info.endpoints.add(endpoint);
-        info.requestQueue.addFirst(endpoint);
+        info.requestQueue.addFirst(endpoint); // TODO not sure if it is correct - given we've just marked this schema version as received, why do we add a request to receive it?
```

Comment:
Still trying to understand why it's needed, but for sure a quick attempt at removing it breaks the tests.

Reply:
The same logic is in pullComplete.
Seems to me we mark before we deal with it in maybePullSchema, I am still not sure why though.... looking again into that

Reply:
Actually, it hasn't necessarily marked are received, has it? It's only marked as received if it's equals to the local schema, I think.

Reply:
Yes, reading back that is right but then it is confusing why in those cases we add to the requestQueue which seems to be a queue of endpoints from which we are going to fetch.... do we need still to fetch? :-)

Reply:
Oh, that is covered in maybePullSchema, we check whether it was already marked received on the first line... It is just not immediately clear before you look into maybePullSchema... perhaps we can simplify or at least comment a bit...

Reply:
At that point we  do not remove the endpoint from the queue, we seem to remove it next time we call maybePullSchema an we iterate the queue. Not sure why that approach was taken really and not just not adding it at all in case we markReceived... I will experiment

Reply:
So we add it to `requestQueue` because we might need to fetch it, unless it has been marked as received. I guess the simplified version would be adding the endpoint to the request queue only if hasn't been marked, something like:
```java
VersionInfo info = versionInfo.computeIfAbsent(version, VersionInfo::new);
info.endpoints.add(endpoint);
logger.trace("Added endpoint {} to schema {}: {}", endpoint, info.version, info);

if (Objects.equals(schemaVersion.get(), version))
{
    info.markReceived();
    logger.trace("Schema {} from {} has been marked as received because it is equal the local schema", version, endpoint);
}
else
{
    info.requestQueue.addFirst(endpoint);
}
```

Reply:
At least the `MigrationCoordinatorTest` and `SchemaTest` in-jvm and unit test pass for me locally with this change

Reply:
exactly, that was what I've commented - "...given we've just marked this schema version as received..." :-)